In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# 이거 실행하고 다시 런타임 시작
# 한글 폰트
!sudo apt-get install -y fonts-nanum
!sudo fc-cache -fv
!rm ~/.cache/matplotlib -rf

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'sudo apt autoremove' to remove it.
The following NEW packages will be installed:
  fonts-nanum
0 upgraded, 1 newly installed, 0 to remove and 27 not upgraded.
Need to get 9,604 kB of archives.
After this operation, 29.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 fonts-nanum all 20170925-1 [9,604 kB]
Fetched 9,604 kB in 1s (9,939 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletyp

In [4]:
import pandas as pd
import numpy as np
#from pykrx import stock
import re
pd.options.display.max_columns = None

In [6]:
df = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-2021_원본.csv')

# 2021년 추가
df2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/재무제표2021.csv')

In [7]:
df.head(10)

Massive Download Unnamed: 1 Unnamed: 2     Unnamed: 3      Unnamed: 4  \
0                 NaN        NaN        NaN            NaN             NaN   
1  Portfolio - kosdaq        NaN        NaN            NaN             NaN   
2                 NaN        NaN        NaN            NaN             NaN   
3                 NaN        NaN        NaN            NaN             NaN   
4                 NaN        NaN        NaN    2018/Annual     2018/Annual   
5                 KIS      Stock       Name  총자산증가율/091010  자기자본증가율/091050   
6              036644     389500  (주)에스비비테크          46.03          -39.46   
7              PM6283     441270    (주)파인엠텍            NaN             NaN   
8              497730     296640    (주)이노룰스          81.84          215.02   
9              499459     417970   모델솔루션(주)          23.53           44.46   

      Unnamed: 5      Unnamed: 6             Unnamed: 7     Unnamed: 8  \
0            NaN             NaN                    NaN            NaN   
1            NaN             NaN                    NaN            NaN   
2            NaN             NaN                    NaN            NaN   
3            NaN             NaN                    NaN            NaN   
4    2018/Annual     2018/Annual            2018/Annual    2018/Annual   
5  매출액증가율/091060  영업이익증가율/091065  법인세비용차감전순이익증가율/091070  순이익증가율/091080   
6           2.26             NaN                    NaN            NaN   
7            NaN             NaN                    NaN            NaN   
8          -6.23          -67.95                 -81.56         -76.96   
9           9.65           73.99                 135.39         136.48   

       Unnamed: 9      Unnamed: 10             Unnamed: 11     Unnamed: 12  \
0             NaN              NaN                     NaN             NaN   
1             NaN              NaN                     NaN             NaN   
2             NaN              NaN                     NaN             NaN   
3             NaN              NaN                     NaN             NaN   
4     2018/Annual      2018/Annual             2018/Annual     2018/Annual   
5  종업원수증가율/091090  총자본영업이익율/092005  총자본법인세비용차감전순이익율/092010  총자본순이익율/092020   
6           16.88            -5.23                   -9.21           -9.21   
7             NaN              NaN                     NaN             NaN   
8             NaN             3.88                    1.19            1.43   
9            4.30            25.30                   24.73           19.34   

               Unnamed: 13      Unnamed: 14             Unnamed: 15  \
0                      NaN              NaN                     NaN   
1                      NaN              NaN                     NaN   
2                      NaN              NaN                     NaN   
3                      NaN              NaN                     NaN   
4              2018/Annual      2018/Annual             2018/Annual   
5  자기자본법인세비용차감전순이익율/092060  자기자본순이익율/092070  자본금법인세비용차감전순이익율/092080   
6                   -28.93           -28.92                 -168.79   
7                      NaN              NaN                     NaN   
8                     3.43             4.11                   57.29   
9                    55.15            43.14                  982.56   

      Unnamed: 16             Unnamed: 17     Unnamed: 18      Unnamed: 19  \
0             NaN                     NaN             NaN              NaN   
1             NaN                     NaN             NaN              NaN   
2             NaN                     NaN             NaN              NaN   
3             NaN                     NaN             NaN              NaN   
4     2018/Annual             2018/Annual     2018/Annual      2018/Annual   
5  자본금순이익율/092090  매출액법인세비용차감전순이익율/092100  매출액순이익율/092110  매출액영업이익율/092120   
6         -168.76                  -20.80          -20.80           -11.80   
7             NaN                     NaN             NaN         

In [8]:
df2.head(10)

Massive Download Unnamed: 1 Unnamed: 2     Unnamed: 3      Unnamed: 4  \
0                 NaN        NaN        NaN            NaN             NaN   
1  Portfolio - kosdaq        NaN        NaN            NaN             NaN   
2                 NaN        NaN        NaN            NaN             NaN   
3                 NaN        NaN        NaN            NaN             NaN   
4                 NaN        NaN        NaN    2020/Annual     2020/Annual   
5                 KIS      Stock       Name  총자산증가율/091010  자기자본증가율/091050   
6              036644     389500  (주)에스비비테크         -23.07          747.37   
7              PM6283     441270    (주)파인엠텍            NaN             NaN   
8              497730     296640    (주)이노룰스         -17.63          -18.42   
9              499459     417970   모델솔루션(주)          -6.52           -7.50   

      Unnamed: 5 Report as of : 2022-10-28      Unnamed: 7      Unnamed: 8  \
0            NaN                       NaN             NaN             NaN   
1            NaN                       NaN             NaN             NaN   
2            NaN                       KRW             NaN             NaN   
3            NaN                       NaN             NaN             NaN   
4    2020/Annual               2020/Annual     2020/Annual     2020/Annual   
5  매출액증가율/091060           총자본영업이익율/092005  총자본순이익율/092020  자본금순이익율/092090   
6           9.76                    -15.18          -19.01         -229.83   
7            NaN                       NaN             NaN             NaN   
8         -20.68                      8.25           11.05          519.76   
9          -9.16                     15.68            2.97          138.47   

       Unnamed: 9      Unnamed: 10    Unnamed: 11  Unnamed: 12  \
0             NaN              NaN            NaN          NaN   
1             NaN              NaN            NaN          NaN   
2             NaN              NaN            NaN          NaN   
3             NaN              NaN            NaN          NaN   
4     2020/Annual      2020/Annual    2020/Annual  2020/Annual   
5  매출액순이익율/092110  매출액영업이익율/092120  자기자본비율/093010  부채비율/093060   
6          -49.59           -39.60          11.87       742.79   
7             NaN              NaN            NaN          NaN   
8           14.96            11.17          54.46        83.64   
9            2.36            12.49          53.74        86.09   

         Unnamed: 13    Unnamed: 14     Unnamed: 15       Unnamed: 16  \
0                NaN            NaN             NaN               NaN   
1                NaN            NaN             NaN               NaN   
2                NaN            NaN             NaN               NaN   
3                NaN            NaN             NaN               NaN   
4        2020/Annual    2020/Annual     2020/Annual       2020/Annual   
5  유보액/납입자본비율/093190  총자본회전율/094010  자기자본회전율/094020  당기순이익(손실)/029000   
6               3.70           0.38            6.64   -3,276,516,000.   
7                NaN            NaN             NaN               NaN   
8           3,253.12           0.74            1.35    1,591,312,000.   
9           2,324.91           1.26            2.32    1,299,209,000.   

           Unnamed: 17    Unnamed: 18     Unnamed: 19    Unnamed: 20  \
0                  NaN            NaN             NaN            NaN   
1                  NaN            NaN             NaN            NaN   
2                  NaN            NaN             NaN            NaN   
3                  NaN            NaN             NaN            NaN   
4          2020/Annual    2021/Annual     2021/Annual    2021/Annual   
5  영업활동으로인한현금흐름/061000  총자산증가율/091010  자기자본증가율/091050  매출액증가율/091060   
6      -1,165,070,000.          17.62          270.81           2.61   
7                  NaN            NaN             NaN            NaN   
8       1,613,872,000.          34.80           84.24          50.27   
9       8,304,769,000.           14.

In [24]:
# 2021년 추가
df2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/재무제표2021.csv')


def make_preprocessing_fs_2021(df):

 

  col_name = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    col_name.append(f"{df[i].iloc[5]} ({df[i].iloc[4].split('/')[0]})")

  df.columns = col_name
  df.drop([0,1,2,3,4,5], axis=0, inplace=True)
  df = df.reset_index(drop=True)


  # 뒤에 널값인 데이터도 같이 들어왔는데 같이 삭제
  last_index = df[df['Name']=='(주)아바텍'].index # 데이터가 들어있는 맨마지막 인덱스. 이 이후꺼를 삭제시키면 된다.
  df = df.iloc[:list(df[df['Name']=='(주)아바텍'].index).pop()+1, :]
  
  
  # 모든 재무제표가 비어있는 코드 drop
  df = df.T
  fs_all_null_code = [] # 모든 재무제표가 null값인 코드 저장을 위해
  for i in df.columns:
    if df[i].isna().sum()==len(df)-3: # KIS, Stock, Name을 제외한 모든 재무제표 컬럼의 개수
      fs_all_null_code.append(i)
  df.drop(fs_all_null_code, axis=1, inplace=True)
  df


  df = df.T # 다시 돌리기
  df = df.reset_index(drop = True)

  
  # stock code 없는 '에스케이제5호기업인수목적(주)' 삭제
  del_index = list(df[df['Stock'].isnull()].index)
  df.drop(del_index, axis=0,inplace=True)
  df= df.reset_index(drop= True)


  # 인수합병된 기업들 빼기
  df['인수합병여부'] = 0
  df['인수합병여부'] = df.apply(lambda x: 1 if '인수목적' in x['Name'] else 0, axis=1)


  df = df[df['인수합병여부']==0]
  df.drop('인수합병여부', axis=1, inplace=True)
  df = df.reset_index(drop = True)




 
  
  

  

  # 컬럼명 다시 깔끔하게 만들기
  col_new = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    start_index = i.find('/')
    col_new.append(i[:start_index] + i[start_index+7:])
  df.columns = col_new
  df.rename(columns={'유보액/093190 (2018)':'유보액/납입자본비율 (2018)', '유보액/093190 (2019)':'유보액/납입자본비율 (2019)', '유보액/093190 (2020)':'유보액/납입자본비율 (2020)', '유보액/093190 (2021)':'유보액/납입자본비율 (2021)'}, inplace = True)


  # 문자열 float로 변경
  for i in df.columns[3:]:
    df[i] = df[i].astype('str')
    df[i] = df[i].apply(lambda x: re.sub(",","",x))
    df[i] = df[i].astype('float')


  # # feature selection = 2021년 재무제표 데이터만
  df = df[['Stock','총자산증가율 (2021)',
       '자기자본증가율 (2021)', '매출액증가율 (2021)', '총자본영업이익율 (2021)', '총자본순이익율 (2021)',
       '자본금순이익율 (2021)', '매출액순이익율 (2021)', '매출액영업이익율 (2021)', '자기자본비율 (2021)',
       '부채비율 (2021)', '유보액/납입자본비율 (2021)', '총자본회전율 (2021)', '자기자본회전율 (2021)',
       '당기순이익(손실) (2021)', '영업활동으로인한현금흐름 (2021)']]
  
  return df

  
df2 = make_preprocessing_fs_2021(df2)

In [25]:
df2.columns

Index(['Stock', '총자산증가율 (2021)', '자기자본증가율 (2021)', '매출액증가율 (2021)',
       '총자본영업이익율 (2021)', '총자본순이익율 (2021)', '자본금순이익율 (2021)', '매출액순이익율 (2021)',
       '매출액영업이익율 (2021)', '자기자본비율 (2021)', '부채비율 (2021)', '유보액/납입자본비율 (2021)',
       '총자본회전율 (2021)', '자기자본회전율 (2021)', '당기순이익(손실) (2021)',
       '영업활동으로인한현금흐름 (2021)'],
      dtype='object')

In [26]:
df2.head(3)

Stock  총자산증가율 (2021)  자기자본증가율 (2021)  매출액증가율 (2021)  총자본영업이익율 (2021)  \
0  389500          17.62          270.81           2.61           -13.65   
1  296640          34.80           84.24          50.27            21.71   
2  417970          14.00           18.82           9.20            18.82   

   총자본순이익율 (2021)  자본금순이익율 (2021)  매출액순이익율 (2021)  매출액영업이익율 (2021)  \
0          -17.26         -138.70          -41.54           -32.85   
1           19.71          269.93           18.83            20.74   
2           15.64          391.32           11.79            14.19   

   자기자본비율 (2021)  부채비율 (2021)  유보액/납입자본비율 (2021)  총자본회전율 (2021)  \
0          37.41       167.32             181.23           0.42   
1          74.43        34.35             737.77           1.05   
2          56.01        78.54             907.47           1.33   

   자기자본회전율 (2021)  당기순이익(손실) (2021)  영업활동으로인한현금흐름 (2021)  
0            1.62     -2.815924e+09        -1.715607e+09  
1            1.59      3.009484e+09         2.473444e+09  
2            2.41      7.086006e+09         9.397543e+09

In [20]:
# 재무제표 가공
df = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-2021_원본.csv')


def make_preprocessing_fs(df):

  # col_name = ['KIS', 'Stock', 'Name']
  # # kisvalue 데이터 다루기 편하게 handling
  # for i in df.columns[3:]:
  #   col_name.append(f"{df[i].iloc[5].split('/')[0]} ({df[i].iloc[4].split('/')[0]})")

  col_name = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    col_name.append(f"{df[i].iloc[5]} ({df[i].iloc[4].split('/')[0]})")

  df.columns = col_name
  df.drop([0,1,2,3,4,5], axis=0, inplace=True)
  df = df.reset_index(drop=True)


  # 뒤에 널값인 데이터도 같이 들어왔는데 같이 삭제
  last_index = df[df['Name']=='(주)아바텍'].index # 데이터가 들어있는 맨마지막 인덱스. 이 이후꺼를 삭제시키면 된다.
  df = df.iloc[:list(df[df['Name']=='(주)아바텍'].index).pop()+1, :]
  
  
  # 모든 재무제표가 비어있는 코드 drop
  df = df.T
  fs_all_null_code = [] # 모든 재무제표가 null값인 코드 저장을 위해
  for i in df.columns:
    if df[i].isna().sum()==len(df)-3: # KIS, Stock, Name을 제외한 모든 재무제표 컬럼의 개수
      fs_all_null_code.append(i)
  df.drop(fs_all_null_code, axis=1, inplace=True)
  df


  df = df.T # 다시 돌리기
  df = df.reset_index(drop = True)

  
  # stock code 없는 '에스케이제5호기업인수목적(주)' 삭제
  del_index = list(df[df['Stock'].isnull()].index)
  df.drop(del_index, axis=0,inplace=True)
  df= df.reset_index(drop= True)


  # 인수합병된 기업들 빼기
  df['인수합병여부'] = 0
  df['인수합병여부'] = df.apply(lambda x: 1 if '인수목적' in x['Name'] else 0, axis=1)


  df = df[df['인수합병여부']==0]
  df.drop('인수합병여부', axis=1, inplace=True)
  df = df.reset_index(drop = True)




  # 컬럼 selection ->삭제할거 삭제하기
  del_list = ['영업이익증가율', '비용차감전순이익증가율', '순이익증가율', '종업원수증가율', '총자본법인세비용차감전순이익율', '자기자본법인세비용차감전순이익율', '자기자본순이익율', '자본금법인세비용차감전순이익율', '매출액법인세비용차감전순이익율', '매출액(종업원1인당)', '법인세비용차감전순이익(종업원1인당)','순이익(종업원1인당)', '인건비(종업원1인당)' , '평균발행주식수', '유형,임대주택자산감가상각비','무형자산상각비', '매출액/121000', '무형자산(계)', '자본총계/118900', '[보통주발행주식수]', '[우선주배당금]' , '보통주배당금', '영업이익(손실)',  '유형고정자산증가율', '유동자산증가율', '재고자산증가율', '2021']

  drop_col_list = []
  for i in del_list:
    for j in df.columns:
      if i in j:
        drop_col_list.append(j)
  

  df.drop(drop_col_list,axis=1,inplace=True)

  # 컬럼명 다시 깔끔하게 만들기
  col_new = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    start_index = i.find('/')
    col_new.append(i[:start_index] + i[start_index+7:])
  df.columns = col_new
  df.rename(columns={'유보액/093190 (2018)':'유보액/납입자본비율 (2018)', '유보액/093190 (2019)':'유보액/납입자본비율 (2019)', '유보액/093190 (2020)':'유보액/납입자본비율 (2020)'}, inplace = True)


  # 문자열 float로 변경
  for i in df.columns[3:]:
    df[i] = df[i].astype('str')
    df[i] = df[i].apply(lambda x: re.sub(",","",x))
    df[i] = df[i].astype('float')

  
  return df

  
df = make_preprocessing_fs(df)

In [27]:
# 2021년 재무제표 데이터 중에서 원본 재무제표 데이터에 없는 종목코드
for i in df2['Stock'].unique():
  if i not in df['Stock'].unique():
    print(i)

391710
383310
335870


In [28]:
# 원본 재무제표 데이터 중에서 2021년 재무제표 데이터에 없는 종목코드는 없다.
for i in df['Stock'].unique():
  if i not in df2['Stock'].unique():
    print(i)

In [29]:
print(df.shape)
print(df2.shape)

(1430, 48)
(1433, 16)


In [30]:
# 종목코드 맞는 것만 merge
final_df = pd.merge(df,df2, on = 'Stock', how='left')
final_df.shape

(1430, 63)

In [31]:
final_df.head(5)

KIS   Stock       Name  총자산증가율 (2018)  자기자본증가율 (2018)  매출액증가율 (2018)  \
0  036644  389500  (주)에스비비테크          46.03          -39.46           2.26   
1  497730  296640    (주)이노룰스          81.84          215.02          -6.23   
2  499459  417970   모델솔루션(주)          23.53           44.46           9.65   
3  HV2035  393890   더블유씨피(주)          69.90           68.28            NaN   
4  HS8995  314140   (주)알피바이오          38.75           31.69          -5.02   

   총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  매출액순이익율 (2018)  \
0            -5.23           -9.21         -168.76          -20.80   
1             3.88            1.43           68.58            1.90   
2            25.30           19.34          768.48           12.97   
3            -8.72           -7.59         -559.62             NaN   
4             8.28            5.66          146.00            6.74   

   매출액영업이익율 (2018)  자기자본비율 (2018)  부채비율 (2018)  유보액/납입자본비율 (2018)  \
0           -11.80          20.22       394.44             340.07   
1             5.16          40.95       144.17            2371.70   
2            16.97          47.94       108.60            2005.45   
3              NaN          93.30         7.18            8560.18   
4             9.85          58.82        70.00            1558.22   

   총자본회전율 (2018)  자기자본회전율 (2018)  당기순이익(손실) (2018)  영업활동으로인한현금흐름 (2018)  \
0           0.44            1.39     -1.917031e+09         6.075800e+07   
1           0.75            2.16      1.770280e+08         1.148444e+09   
2           1.49            3.33      7.210181e+09         5.469434e+09   
3            NaN             NaN     -9.513513e+09        -8.218473e+09   
4           0.84            1.40      4.363446e+09         6.995499e+09   

   총자산증가율 (2019)  자기자본증가율 (2019)  매출액증가율 (2019)  총자본영업이익율 (2019)  \
0         -21.16          -95.80         -34.70           -17.06   
1          -1.20           32.63          43.95            17.43   
2           9.91           24.51           8.93            21.58   
3         172.97           31.73            NaN            -1.72   
4          29.38            7.64           7.86             3.85   

   총자본순이익율 (2019)  자본금순이익율 (2019)  매출액순이익율 (2019)  매출액영업이익율 (2019)  \
0          -21.67         -421.59          -79.56           -62.63   
1            3.89          210.40            4.61            20.65   
2           16.81          774.92           12.00            15.41   
3           -2.21         -244.88          -18.71           -14.52   
4            3.33          105.70            4.90             5.67   

   자기자본비율 (2019)  부채비율 (2019)  유보액/납입자본비율 (2019)  총자본회전율 (2019)  \
0           1.08      9183.55             -81.52           0.27   
1          54.98        81.89            2891.58           0.84   
2          54.31        84.14            2521.46           1.40   
3          45.03       122.10            5257.56           0.12   
4          48.94       104.35            1624.67           0.68   

   자기자본회전율 (2019)  당기순이익(손실) (2019)  영업활동으로인한현금흐름 (2019)  총자산증가율 (2020)  \
0            2.31     -4.789125e+09        -2.686112e+09         -23.07   
1            1.76      6.186920e+08         4.188603e+09         -17.63   
2            2.73      7.270528e+09         1.075707e+10          -6.52   
3            0.20     -6.513745e+09        -2.835368e+10          21.46   
4            1.28      3.423335e+09         6.582410e+09          12.53   

   자기자본증가율 (2020)  매출액증가율 (2020)  총자본영업이익율 (2020)  총자본순이익율 (2020)  \
0          747.37           9.76           -15.18          -19.01   
1          -18.42         -20.68             8.25           11.05   
2           -7.50          -9.16            15.68            2.97   
3          -15.77         221.41             2.05           -6.54   
4            7.59          40.18             3.37            3.75   

   자본금순이익율 (2020)  매출액순이익율 (2020)  매출액영업이익율 (2020)  자기자본비율 (2020)  \
0         -229.83          -49.59           -39.60          11.87   
1          51

In [33]:
final_df.isna().sum()[:20]

KIS                     0
Stock                   0
Name                    0
총자산증가율 (2018)          45
자기자본증가율 (2018)         72
매출액증가율 (2018)          63
총자본영업이익율 (2018)        30
총자본순이익율 (2018)         29
자본금순이익율 (2018)         29
매출액순이익율 (2018)         46
매출액영업이익율 (2018)        46
자기자본비율 (2018)          48
부채비율 (2018)            48
유보액/납입자본비율 (2018)      29
총자본회전율 (2018)          55
자기자본회전율 (2018)         71
당기순이익(손실) (2018)       52
영업활동으로인한현금흐름 (2018)    62
총자산증가율 (2019)          31
자기자본증가율 (2019)         66
dtype: int64

In [37]:
df = final_df

# 관리종목 연도별 종목코드 뽑기


In [34]:
# 2018, 2019, 2020, 2021
df_2018_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-1_수정.csv')
df_2018_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-2_수정.csv')

df_2019_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-1_수정.csv')
df_2019_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-2_수정.csv')
df_2019_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-3_수정.csv')

df_2020_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-1_수정.csv')
df_2020_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-2_수정.csv')
df_2020_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-3_수정.csv')

df_2021_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-1_수정.csv')
df_2021_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-2_수정.csv')
df_2021_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-3_수정.csv')

In [35]:
# 관리대상기업 데이터 연도별 concat
df_2018 = pd.concat([df_2018_1, df_2018_2], axis=0).reset_index(drop=True)
df_2019 = pd.concat([df_2019_1, df_2019_2, df_2019_3], axis=0).reset_index(drop=True)
df_2020 = pd.concat([df_2020_1, df_2020_2, df_2020_3], axis=0).reset_index(drop=True)
df_2021 = pd.concat([df_2021_1, df_2021_2, df_2021_3], axis=0).reset_index(drop=True)


In [36]:
# 관리종목 지정된 종목코드 리스트 반환하는 함수
def return_risk_code(df):
  df['종목코드'] = df['종목코드'].astype('str')
  df['종목코드'] = df['종목코드'].apply(lambda x: x.zfill(6) if len(x)!=6 else x) # 종목코드 6자리로 만들어주기

  df['지정/해제여부'] = df.apply(lambda x: 0 if '해제' in x['공시제목'] else 1, axis=1)
  df = df[df['지정/해제여부']==1].reset_index(drop=True) # 관리종목 지정인 데이터만 불러오기
  
  종목코드_list = list(df['종목코드'].unique())
  return 종목코드_list


risk_code_2018 = return_risk_code(df_2018)
risk_code_2019 = return_risk_code(df_2019)
risk_code_2020 = return_risk_code(df_2020)
risk_code_2021 = return_risk_code(df_2021)

print(f'2018년도 관리종목기업코드 개수 : {len(risk_code_2018)}')
print(f'2019년도 관리종목기업코드 개수 : {len(risk_code_2019)}')
print(f'2020년도 관리종목기업코드 개수 : {len(risk_code_2020)}')
print(f'2021년도 관리종목기업코드 개수 : {len(risk_code_2021)}')


2018년도 관리종목기업코드 개수 : 75
2019년도 관리종목기업코드 개수 : 93
2020년도 관리종목기업코드 개수 : 101
2021년도 관리종목기업코드 개수 : 98


# 기존 KisValue데이터에 관리종목여부 추가
- 2019, 2020, 2021 여부만 추가할 것.
- 모델링할 때 데이터 불균형 문제를 어느정도 해결해야할 것 같다.

In [39]:
# 일단 모든 값 0으로 초기화
# 인수합병된 기업들을 미리 지워서 그런지 위에서 출력한 개수보다 적게 나온다.


df['2018_관리종목여부'] = 0
df['2019_관리종목여부'] = 0
df['2020_관리종목여부'] = 0
df['2021_관리종목여부'] = 0

df['2018_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2018 else 0 , axis=1)
df['2019_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2019 else 0 , axis=1)
df['2020_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2020 else 0 , axis=1)
df['2021_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2021 else 0 , axis=1)

print(df['2018_관리종목여부'].value_counts())
print(df['2019_관리종목여부'].value_counts())
print(df['2020_관리종목여부'].value_counts())
print(df['2021_관리종목여부'].value_counts())

0    1407
1      23
Name: 2018_관리종목여부, dtype: int64
0    1388
1      42
Name: 2019_관리종목여부, dtype: int64
0    1392
1      38
Name: 2020_관리종목여부, dtype: int64
0    1400
1      30
Name: 2021_관리종목여부, dtype: int64


In [40]:
df.describe()

총자산증가율 (2018)  자기자본증가율 (2018)  매출액증가율 (2018)  총자본영업이익율 (2018)  \
count   1.385000e+03     1358.000000    1367.000000      1400.000000   
mean    8.307525e+02      685.906127      21.573921         2.502879   
std     2.926893e+04    23315.797840     153.383154        15.853033   
min    -6.804000e+01      -90.030000    -100.000000      -129.530000   
25%    -1.530000e+00       -1.100000      -8.145000        -0.775000   
50%     7.340000e+00        6.950000       4.380000         2.905000   
75%     2.496000e+01       23.480000      20.830000         8.742500   
max     1.089023e+06   858889.270000    3648.630000       137.560000   

       총자본순이익율 (2018)  자본금순이익율 (2018)  매출액순이익율 (2018)  매출액영업이익율 (2018)  \
count     1401.000000     1401.000000     1384.000000      1384.000000   
mean        -1.536881       70.205332     -616.313439      -471.265383   
std         26.836112     2323.153378    11080.499725     10273.463568   
min       -532.050000   -64738.350000  -353205.520000   -350051.850000   
25%         -3.400000      -31.970000       -4.920000        -0.890000   
50%          2.620000       33.020000        3.670000         4.455000   
75%          7.760000      143.810000       10.500000        11.710000   
max        114.400000    52574.620000     1307.340000        77.620000   

       자기자본비율 (2018)  부채비율 (2018)  유보액/납입자본비율 (2018)  총자본회전율 (2018)  \
count    1382.000000  1382.000000        1401.000000    1375.000000   
mean       64.496852    97.685753        1697.744404       0.802269   
std        20.955324   299.279136        9641.966500       0.633225   
min         1.120000     0.910000       -8327.170000       0.010000   
25%        49.792500    21.345000         334.930000       0.400000   
50%        65.125000    53.550000         783.650000       0.660000   
75%        82.407500   100.827500        1592.300000       1.050000   
max        99.100000  8799.540000      248880.480000       8.440000   

       자기자본회전율 (2018)  당기순이익(손실) (2018)  영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  \
count     1359.000000      1.378000e+03         1.368000e+03    1399.000000   
mean         1.610412      2.820362e+09         5.502024e+09      21.805661   
std          2.491764      1.698619e+10         1.958206e+10      77.475601   
min          0.010000     -1.234088e+11        -1.672801e+11     -82.480000   
25%          0.580000     -2.380625e+09        -1.445045e+09      -1.220000   
50%          1.050000      1.959088e+09         2.313819e+09       7.260000   
75%          1.880000      6.327230e+09         8.111196e+09      22.825000   
max         60.310000      1.546099e+11         2.871206e+11    1628.770000   

       자기자본증가율 (2019)  매출액증가율 (2019)  총자본영업이익율 (2019)  총자본순이익율 (2019)  \
count     1364.000000    1381.000000      1419.000000     1418.000000   
mean        32.059802     116.543526         1.872854       -2.697045   
std        173.069248    3535.374771        16.070914       25.882170   
min        -98.300000    -100.000000      -191.540000     -353.270000   
25%         -1.172500      -8.560000        -0.425000       -3.970000   
50%          5.500000       5.300000         2.810000        2.200000   
75%         19.890000      21.100000         8.230000        7.217500   
max       3408.850000  131338.880000        89.070000       71.040000   

       자본금순이익율 (2019)  매출액순이익율 (2019)  매출액영업이익율 (2019)  자기자본비율 (2019)  \
count     1420.000000     1399.000000      1400.000000    1400.000000   
mean      -437.646775     -369.679006      -110.732171      64.143086   
std      19593.758488     6471.426055      1343.674254      20.981897   
min    -734161.410000  -184682.910000    -32961.540000       1.080000   
25%        -35.380000       -7.115000        -0.497500      49.122500   
50%         27.930000        3.100000         4.490000      65.405000   
75%        126.837500       10.230000        11.542500      82.162500   
max      68510.460000     1085.710000        99.850000      99.570000   

       부채비율 (2019)  유보액

In [41]:
# 2019 관리종목대상의 2018 재무제표 평균 보기
print(df[df['2019_관리종목여부']==1]['총자산증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자본금순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['부채비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['유보액/납입자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['당기순이익(손실) (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['영업활동으로인한현금흐름 (2018)'].mean())

8.886904761904763
7.115714285714287
-3.3880952380952385
-8.333809523809524
-17.688333333333333
-106.68761904761905
-51.57119047619048
-26.022857142857145
49.95333333333333
142.62261904761905
564.6380952380953
0.5435714285714286
1.2719047619047619
-10126616951.219513
-1637915292.682927


In [42]:
# 위랑 비교해서 2019 관리종목대상의 2018 재무제표 평균 보기
print(df[df['2019_관리종목여부']==0]['총자산증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['매출액증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자본금순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['매출액순이익율 (2018)'].mean()) # 이거 아웃라이어 있는거같은데 살짝 지우자. 
print(df[df['2019_관리종목여부']==0]['매출액영업이익율 (2018)'].mean()) # 이것도 아웃라이어 데이터 삭제해주자.
print(df[df['2019_관리종목여부']==0]['자기자본비율 (2018)'].mean())

print(df[df['2019_관리종목여부']==0]['부채비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['유보액/납입자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['영업활동으로인한현금흐름 (2018)'].mean())

856.4549218168281
707.5696504559271
22.365169811320758
2.838033873343152
-1.037718910963944
75.67222222222223
-633.9879359165425
-485.19994783904616
64.95269402985075
96.27728358208955
64.95269402985075
1732.7631420161883
0.8104201050262565
1.6212072892938498
5722624618.688771


## 일단 2018년 컬럼 모두가 null이면 drop(axis=0) 하자
- 2017년이나 2018년 이전부터 설립된 기업만을 대상으로 하기 위해
- 내가 확인한결과 2020, 2021은 모두 nan인게 없다.

In [43]:
df3 = df.iloc[:,3:18].T
all_nan_2018 = list(df3.isna().sum()[df3.isna().sum().values==15].index)

df = df.drop(all_nan_2018, axis=0).reset_index(drop=True)

In [44]:
print(df['2018_관리종목여부'].value_counts())
print(df['2019_관리종목여부'].value_counts())
print(df['2020_관리종목여부'].value_counts())
print(df['2021_관리종목여부'].value_counts())

0    1378
1      23
Name: 2018_관리종목여부, dtype: int64
0    1359
1      42
Name: 2019_관리종목여부, dtype: int64
0    1363
1      38
Name: 2020_관리종목여부, dtype: int64
0    1372
1      29
Name: 2021_관리종목여부, dtype: int64


In [49]:
df

KIS   Stock           Name  총자산증가율 (2018)  자기자본증가율 (2018)  \
0     036644  389500      (주)에스비비테크          46.03          -39.46   
1     497730  296640        (주)이노룰스          81.84          215.02   
2     499459  417970       모델솔루션(주)          23.53           44.46   
3     HV2035  393890       더블유씨피(주)          69.90           68.28   
4     HS8995  314140       (주)알피바이오          38.75           31.69   
...      ...     ...            ...            ...             ...   
1396  I73639  950110  SBI핀테크솔루션즈(주)         146.31           77.94   
1397  125028  046970         (주)우리로          42.14            8.83   
1398  383046  113810         디젠스(주)         -12.94            6.17   
1399  F11411  155650        와이엠씨(주)         -22.34           13.57   
1400  099145  149950         (주)아바텍          -2.81           -0.35   

      매출액증가율 (2018)  총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  \
0              2.26            -5.23           -9.21         -168.76   
1             -6.23             3.88            1.43           68.58   
2              9.65            25.30           19.34          768.48   
3               NaN            -8.72           -7.59         -559.62   
4             -5.02             8.28            5.66          146.00   
...             ...              ...             ...             ...   
1396          10.83             8.05            7.49           24.84   
1397          92.45             4.43           -0.55           -2.82   
1398         -29.81            -3.98          -16.02          -87.40   
1399          59.47            10.09            7.46          196.38   
1400         -22.27             3.76            5.22           94.76   

      매출액순이익율 (2018)  매출액영업이익율 (2018)  자기자본비율 (2018)  부채비율 (2018)  \
0             -20.80           -11.80          20.22       394.44   
1               1.90             5.16          40.95       144.17   
2              12.97            16.97          47.94       108.60   
3                NaN              NaN          93.30         7.18   
4               6.74             9.85          58.82        70.00   
...              ...              ...            ...          ...   
1396           16.69            17.92          45.68       118.91   
1397           -1.06             8.50          48.09       107.95   
1398          -31.83            -7.91          38.61       158.99   
1399            6.33             8.55          50.08        99.67   
1400           10.39             7.49          93.86         6.55   

      유보액/납입자본비율 (2018)  총자본회전율 (2018)  자기자본회전율 (2018)  당기순이익(손실) (2018)  \
0                340.07           0.44            1.39     -1.917031e+09   
1               2371.70           0.75            2.16      1.770280e+08   
2               2005.45           1.49            3.33      7.210181e+09   
3               8560.18            NaN             NaN     -9.513513e+09   
4               1558.22           0.84            1.40      4.363446e+09   
...                 ...            ...             ...               ...   
1396              93.99           0.45            0.88      3.963095e+09   
1397             194.50           0.52            0.96     -3.391840e+08   
1398              63.05           0.50            1.44     -1.185408e+10   
1399             800.95           1.18            2.87      1.464977e+10   
1400            1650.60           0.50            0.54      7.726745e+09   

      영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  자기자본증가율 (2019)  매출액증가율 (2019)  \
0            6.075800e+07         -21.16          -95.80         -34.70   
1            1.148444e+09          -1.20           32.63          43.95   
2            5.469434e+09           9.91           24.51           8.93   
3           -8.218473e+09         172.97           31.73            NaN   
4            6.995499e+09          29.38            7.64           7.86   
...                   ...            ...             ...            ...   
1396         2.682301e+09           

# 최종 파일 만들기
- nan이 있는 종목 drop하기
- nan이 있는 데이터 117개 drop
- 1284개 기업 데이터로 확정

In [54]:

df22 = df.copy()
df22['NaN_cnt'] = df22.isnull().sum(1)
del_list = list(df22[df22['NaN_cnt']>0].index)
print(len(del_list))

122


In [55]:
df22.drop(del_list, axis=0, inplace=True)
df22.isna().sum()

KIS               0
Stock             0
Name              0
총자산증가율 (2018)     0
자기자본증가율 (2018)    0
                 ..
2018_관리종목여부       0
2019_관리종목여부       0
2020_관리종목여부       0
2021_관리종목여부       0
NaN_cnt           0
Length: 68, dtype: int64

In [56]:
df22.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1279 entries, 0 to 1400
Data columns (total 68 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   KIS                  1279 non-null   object 
 1   Stock                1279 non-null   object 
 2   Name                 1279 non-null   object 
 3   총자산증가율 (2018)        1279 non-null   float64
 4   자기자본증가율 (2018)       1279 non-null   float64
 5   매출액증가율 (2018)        1279 non-null   float64
 6   총자본영업이익율 (2018)      1279 non-null   float64
 7   총자본순이익율 (2018)       1279 non-null   float64
 8   자본금순이익율 (2018)       1279 non-null   float64
 9   매출액순이익율 (2018)       1279 non-null   float64
 10  매출액영업이익율 (2018)      1279 non-null   float64
 11  자기자본비율 (2018)        1279 non-null   float64
 12  부채비율 (2018)          1279 non-null   float64
 13  유보액/납입자본비율 (2018)    1279 non-null   float64
 14  총자본회전율 (2018)        1279 non-null   float64
 15  자기자본회전율 (2018)       1279 non-null   f

In [57]:
len(df22)

1279

In [59]:
print(df22['2018_관리종목여부'].value_counts())
print(df22['2019_관리종목여부'].value_counts())
print(df22['2020_관리종목여부'].value_counts())
print(df22['2021_관리종목여부'].value_counts())

0    1258
1      21
Name: 2018_관리종목여부, dtype: int64
0    1240
1      39
Name: 2019_관리종목여부, dtype: int64
0    1248
1      31
Name: 2020_관리종목여부, dtype: int64
0    1254
1      25
Name: 2021_관리종목여부, dtype: int64


In [60]:
# 최종파일 내보내기
df22.reset_index(drop=True, inplace= True)
df22.to_csv('real_real_df.csv', index=False)